pkg/ccl/changefeedccl/schemafeed/timestamp_barrier.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Licensed as a CockroachDB Enterprise file under the Cockroach Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package schemafeed
+
+import (
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/errors"
+)
+
+// A timestampBarrier is used to synchronize goroutines that advance a monotonically
+// increasing frontier timestamp and goroutines that need to wait for the
+// frontier to advance past a specific timestamp.
+//
+// A monotonically decreasing error timestamp can also be tracked and signifies
+// the earliest time at which the frontier cannot advance to/past due to some
+// external error condition. This external error is recorded so that future
+// attempts to wait at a timestamp at or later than the error timestamp will
+// return the error.
+//
+// The data structure itself is not safe for concurrent use and must be
+// protected by a synchronization primitive.
+type timestampBarrier struct {
+	frontier hlc.Timestamp
+	errTS    hlc.Timestamp
+	err      error
+	waiters  []timeWaiter
+}
+
+// A timeWaiter is created for each goroutine that is blocked waiting for the
+// frontier value to advance past a specific timestamp.
+type timeWaiter struct {
+	ts     hlc.Timestamp
+	waitCh chan<- error
+}
+
+// advanceFrontier advances the frontier to the given timestamp.
+// The new frontier value must not be earlier than the current frontier,
+// nor may it be at or later than the current error timestamp.
+func (m *timestampBarrier) advanceFrontier(ts hlc.Timestamp) error {
+	if ts.Less(m.frontier) {
+		return errors.AssertionFailedf(
+			"cannot advance frontier to %s, which is earlier than the current frontier %s", ts, m.frontier)
+	}
+	if m.errTS.IsSet() && m.errTS.LessEq(ts) {
+		return errors.AssertionFailedf(
+			"cannot advance frontier to %s, which is equal to or later than the current error timestamp %s", ts, m.errTS)
+	}
+	m.frontier = ts
+
+	// Signal all waiters that are waiting at or before the new frontier.
+	var newWaiters []timeWaiter
+	for _, w := range m.waiters {
+		if w.ts.LessEq(m.frontier) {
+			close(w.waitCh)
+			continue
+		}
+		newWaiters = append(newWaiters, w)
+	}
+	m.waiters = newWaiters
+
+	return nil
+}
+
+// wait returns a channel that can be used to wait for either the frontier
+// timestamp to reach/exceed the specified timestamp or the error timestamp to
+// recede to/below the specified timestamp, at which point the channel will be
+// closed. In the latter case, the channel will be sent the error associated
+// with the error timestamp before being closed.
+func (m *timestampBarrier) wait(ts hlc.Timestamp) <-chan error {
+	waitCh := make(chan error, 1)
+
+	if ts.LessEq(m.frontier) {
+		close(waitCh)
```

Comment:
This is a common case. That's why I suggested creating a single, closed channel that can be used here and avoid new channel creation in fast path.

Reply:
Ah I missed that, fixed now.